### PR TITLE
Finish shipment confirm and tracking for UPS REST API.

### PR DESCRIPTION
**Pull request by ragnese**
### Description (*)
Attempt at finishing the rest of UPS REST API implementation.

### Related Pull Requests
N/A

### Manual testing scenarios (*)
I've only tested by loading up some test data from our own store and checking that the shipping estimates for a shopping cart match if I switch between the XML and REST implementations.

### Questions or comments
Because Magento2 has the _liveUrls and _defaultUrls while also using fewer config.xml parameters, I thought it made sense for us to merge both approaches and basically use both. This keeps the behavior consistent between the XML and REST modes.

I also updated some of the XML methods to use some of the newly introduced helper methods, since the reason they were implemented in M2 was precisely to be shared between the XML and REST methods that use similar logic.

I tried to add a little bit more error handling around the cURL calls, etc, but I'm not at all sure what the Magento philosophy is on errors and exceptions. So, some guidance there would be appreciated.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All automated tests passed successfully (all builds are green)
 - [x] Add yourself to contributors list

### CODE REVIEW

File: app/code/core/Mage/Usa/Model/Shipping/Carrier/Ups.php [hunk 1]
```diff
@@ -1132,22 +1111,176 @@ protected function _parseXmlTrackingResponse($trackingvalue, $xmlResponse)
             }
         }
 
+        $this->setTrackingResultData($resultArr,$trackingvalue,$errorTitle);
```
Comment (fballiano) [inline app/code/core/Mage/Usa/Model/Shipping/Carrier/Ups.php:1114]:
```suggestion
        $this->setTrackingResultData($resultArr, $trackingvalue, $errorTitle);
```

Comment (fballiano) [reply app/code/core/Mage/Usa/Model/Shipping/Carrier/Ups.php:1114]: phpcs

I can't edit your PR, I'll merge it once this is fixed

**Comment by fballiano** [comment]
@ragnese thank you for this!

could you also check why phpstan is complaining about an undefined method?

**Comment by ragnese** [comment]
One of the things I'd like to ask about is error handling. 

In the UpsAuth class, the `getAccessToken` method throws an exception when cURL fails, and returns `false` if decoding the JSON response fails. Shouldn't both of those be treated as the same class of failure? Shouldn't we either throw an exception in both cases OR return `false` in both cases?

Yet, in the Ups class, whenever we use the `curl_*` functions, we *don't* throw exceptions. Instead, we just add some kind of error information to the return value of the function.

We also don't catch the exception that might be thrown from `UpsAuth->getAccessToken` in the Ups class. Should we?

**Comment by fballiano** [comment]
@ragnese since curl doesn't throw exceptions I guess also in upsauth it would be better to just return false, and in that case the ups model wouldn't have to do anything special right?

**Comment by fballiano** [comment]
> Well, we still need to handle the `false` value, so all calls to get the access token will need _some_ kind of error handling.

mmmm it's probably worth to throw an exception if the token is not valid since it stops the flow.

i'm thinking, if we just return "no shipping rates" the users may not understand that there's a problem with the system

what do you think?

> Does Magento have an overall "philosophy" when it comes to exceptions vs returning error values?

I don't think so, it depends case by case

> After looking over the code again, I think that the methods on the Ups class should _not_ throw exceptions, because they all return these "Result" objects that are able to represent errors/failure, and the classes that call these Carrier methods don't attempt to handle any exceptions. However, I do also think the UpsAuth class probably _should_ throw an exception because if we just return `false`, then we lose the information about what the error actually was.

that is fine by me

> So, I think I'll go back through and make UpsAuth only throw exceptions and not return `false`, and then I'll make sure that all calls to it are just handled properly.

thank you!

**Comment by fballiano** [comment]
@ragnese what if I merge this code already? or do you think you want to take a look at the last things before?

**Comment by ragnese** [comment]
@fballiano, sorry for the delayed response. 

I do think there's room for improvement in how failure is communicated to the end user, but that is probably out of scope for this task. Improving that situation would require analyzing/updating the other carrier implementations as well, and maybe even other classes up the call chain. So, I think this work should mostly be focused on implementing the UPS REST API without any noticeable behavior changes to the end user.

I think the code is pretty much ready to merge. But I mostly tested the rate/quote stuff, and only with a few simple orders with our test data/store. It would probably be a good idea for you to test it with your set up as well, just in case.

**Pull request merged**